version: Show number of commits and commit sha in ZULIP_VERSION.

punchagan opened the pull request:
We use `git describe --tags` to get information about the number of commit since
the last major version, and the sha of the current HEAD. This is added to the
ZULIP_VERSION when a deploy is done from `git`.

Modified heavily by punchagan to:
* to use git describe instead of `git log` and `wc`
* use a separate script to run the git describe command
* write the file with version info to var/ and remove it from the repo

Fixes #4685.

timabbott commented on tools/update-prod-static:
```diff
@@ -34,6 +34,9 @@ os.chdir(settings.DEPLOY_ROOT)
 fp = open('var/log/update-prod-static.log', 'w')
 
+# Setup zulip git version cache file
+run(['./tools/cache-zulip-git-version'], stdout=fp, stderr=fp)
```
I think the Travis CI error is because the way we build release tarballs is I think in a context where we don't have a Git repository by by the time we're running `update-prod-static` (we first make a fresh checkout with `git archive` in that toolchain to avoid issues with untracked files, etc., in the development environment).  So I think perhaps we want to run `./tools/cache-zulip-git-version` in `tools/build-release-tarball` and in `tools/upgrade-zulip-from-git` instead, as the two code paths that call into `tools/update-prod-static`?  I think that'd also avoid confusing results like the value here being cached in a development environment if we ever make `tools/update-prod-static` directly runnable there for debugging.

punchagan replied:
I was actually writing the version file to the `var/` directory, assuming we don't need this in a `release-tarball`. Do we need this information when running from a release tarball?

punchagan replied:
@timabbott Also, do we need this information in a release tarball? Just having the version set when creating the tarball would be good enough, right? If not, what would be a good place to drop the cache file for the version info? Right now I put it in `var/` assuming we only need it when running from `git`.

timabbott replied:
We definitely want it in release tarballs.

I think `var/` is a fine place to put it if that directory ends up being included in a release tarball.  If it doesn't, which seems likely, we could just put it in the root of the repository; I think we have similar other files like `version` in the root of release tarballs.

timabbott commented on version.py:
```diff
@@ -1,4 +1,14 @@
+    with open(zulip_git_version_file) as f:
+        version = f.read().strip()
+        if version:
+            ZULIP_VERSION = version
```
I guess with this implementation, in the development environment it'll always read `2.0.3+git`, but in production it'll have the more expressive value?  I think that's probably the best option, but perhaps we should add a comment explaining that detail?

punchagan replied:
Yeah, that is correct. I'll add a comment. Also, when running from `git` the version reads something like `2.0.0-1213-g6d4c84339` and not `2.0.3-xxxx`. 
I'm not sure if this acceptable, or we want to get the version number differently to fix this.

timabbott replied:
I'm fine with those having a different look; what matters is that it's clear what the version means, and I think that'll be the case in all situations.

timabbott commented:
This is great, thanks @punchagan!  I posted a couple quick comments.

punchagan commented:
@timabbott I've updated this PR based on our discussion. 

- the cache file gets written to the root of the repository and the release tar ball
- I've also added call to it, during provisioning. So, the version should look similar in development to what it would look like in production
- I've also moved the calls to the script creating the version file to `build-release-tarball` and the `upgrade-zulip-stage-2` script.

punchagan commented:
I actually ended up removing the call to `cache-zulip-git-version` during the provisioning step. Sometimes this would end up in showing the wrong version in the development environment, and I guess it's better to just not show it at all.

timabbott commented:
This is exactly what I had in mind; I agree that doing something in provision would just be confusing.  Merged, thanks @punchagan!
